compiler/rustc_mir_transform/src/generator.rs

```diff
@@ -740,24 +740,36 @@ fn sanitize_witness<'tcx>(
         }
     };
 
-    for (local, decl) in body.local_decls.iter_enumerated() {
-        // Ignore locals which are internal or not saved between yields.
-        if !saved_locals.contains(local) || decl.internal {
-            continue;
-        }
-        let decl_ty = tcx.normalize_erasing_regions(param_env, decl.ty);
+    // First check if any of the locals are an uninhabited type. If so, we don't need to do the
+    // rest of this check because this code unreachable.
+    let any_uninhabited = body.local_decls.iter_enumerated().any(|(_, decl)| {
```

Comment:
Since this validates a complete generator, we don't really know which parts of it are unreachable if any.

If we want to optimize out code following call with uninhabited type as unreachable, we should do that by transforming the MIR and ensuring those locals are not saved in the first place, while keeping the validation in place to ensure that typeck and MIR are in sync.